File: leetcode/144. Binary Tree Preorder Traversal/Python3/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorder(self, node, path):
        if node:
            path.append(node.val)
            self.preorder(node.left, path)
            self.preorder(node.right, path)

    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        path = []
        self.preorder(root, path)
        return path

File: leetcode/144. Binary Tree Preorder Traversal/solution.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        if (root == NULL) return res;

        stack<TreeNode*> s { root };

        TreeNode *curr;
        while (!s.empty()) {
            curr = s.top();
            s.pop();
            res.push_back(curr->val);
            if (curr->right) s.push(curr->right);
            if (curr->left) s.push(curr->left);
        }
        return res;
    }
};